HelpersWayland.cpp

```diff
@@ -0,0 +1,295 @@
+    EGL_COLOR_BUFFER_TYPE,     EGL_RGB_BUFFER,
+    EGL_BUFFER_SIZE,           32,
+    EGL_RED_SIZE,              8,
+    EGL_GREEN_SIZE,            8,
+    EGL_BLUE_SIZE,             8,
+    EGL_ALPHA_SIZE,            8,
+
+    EGL_DEPTH_SIZE,            24,
+    EGL_STENCIL_SIZE,          8,
+
+    EGL_SAMPLE_BUFFERS,        0,
+    EGL_SAMPLES,               0,
+
+    EGL_SURFACE_TYPE,          EGL_WINDOW_BIT,
+    EGL_RENDERABLE_TYPE,       EGL_OPENGL_BIT,
+
+    EGL_NONE,
+};
+
+const EGLint egl_context_attribs[] = {
+    EGL_NONE,
+};
+
+const EGLint egl_surface_attribs[] = {
+    EGL_RENDER_BUFFER, EGL_BACK_BUFFER,
+    EGL_NONE,
+};
+
+void HelpersWayland::setup_egl(
+        EGLint api,
+        EGLNativeDisplayType native_display,
+        EGLNativeWindowType native_window,
+        EGLDisplay* out_display,
+        EGLConfig* out_config,
+        EGLContext* out_context,
+        EGLSurface* out_window_surface) {
+
+    EGLint ignore;
+    EGLBoolean ok;
+
+    ok = eglBindAPI(api);
+    if (!ok)
+        error_fatal("eglBindAPI(0x%x) failed", api);
+
+    EGLDisplay display = eglGetDisplay(native_display);
+    if (display == EGL_NO_DISPLAY)
+        error_fatal("eglGetDisplay() failed");
+
+    ok = eglInitialize(display, &ignore, &ignore);
+    if (!ok)
+        error_fatal("eglInitialize() failed");
+
+    EGLint configs_size = 256;
+    EGLConfig* configs = new EGLConfig[configs_size];
+    EGLint num_configs;
+    ok = eglChooseConfig(
+        display,
+        egl_config_attribs,
+        configs,
+        configs_size, // num requested configs
+        &num_configs); // num returned configs
+    if (!ok)
+        error_fatal("eglChooseConfig() failed");
+    if (num_configs == 0)
+        error_fatal("failed to find suitable EGLConfig");
+    EGLConfig config = configs[0];
+    delete [] configs;
+
+    EGLContext context = eglCreateContext(
+        display,
+        config,
+        EGL_NO_CONTEXT,
+        egl_context_attribs);
+    if (!context)
+        error_fatal("eglCreateContext() failed");
+
+    EGLSurface surface = eglCreateWindowSurface(
+        display,
+        config,
+        native_window,
+        egl_surface_attribs);
+    if (!surface)
+        error_fatal("eglCreateWindowSurface() failed");
+
+    ok = eglMakeCurrent(display, surface, surface, context);
+    if (!ok)
+    error_fatal("eglMakeCurrent() failed");
+
+    // Check if surface is double buffered.
+    EGLint render_buffer;
+    ok = eglQueryContext(
+        display,
+        context,
+        EGL_RENDER_BUFFER,
+        &render_buffer);
+    if (!ok)
+        error_fatal("eglQueyContext(EGL_RENDER_BUFFER) failed");
+    if (render_buffer == EGL_SINGLE_BUFFER)
+        printf("warn: EGL surface is single buffered\n");
+
+    *out_display = display;
+    *out_config = config;
+    *out_context = context;
+    *out_window_surface = surface;
+}
+
+const char *HelpersWayland::queryDriverName(int screen) {
+    const char *x_display_name = NULL;
```

Comment:
Why are you mixing xcb calls with egl calls?
As far as I known, you can make everything using egl only.
Is there any reason why xcb is needed?

Reply:
To query driver name or configs,

First, we need to create EGLDisplay, EGLSurface, EGLContext. We are using `eglCreateWindowSurface` to create EGLSurface where we are passing _native_window_ as a argument. To create this window we are using xcb libs.

Reply:
The code is still mixing EGL and X11 calls.
Is there anything that is still missing on X11, or why are you mixing them?

Reply:
Interesting, this makes sense for XWayland.
But in this case we want to specifically **not** use XWayland. So how do you create a EGLSurface in pure Wayland?